### PR TITLE
feat(table): table virtualization

### PR DESCRIPTION
**Pull request by aborjinik**
The ui5-table-virtualizer component is used inside the ui5-table to virtualize the table rows.It is responsible for rendering only the rows that are visible in the viewport and updating them on scroll. This allows large numbers of rows to exist, but maintain high performance by only paying the cost for those that are currently visible.

This component is experimental.

### CODE REVIEW

File: packages/main/src/TableVirtualizer.ts [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+		this._table && this._table._invalidate++;
+	}
+
+	onTableAfterRendering(): void {
+		if (!this._table) {
+			return;
+		}
+
+		this._updateRowsHeight();
+		if (this._tabBlockingState & TabBlocking.Released) {
```
[the remark this answers is not in the document]
Comment (DonkeyCo) [reply packages/main/src/TableVirtualizer.ts:148]: Maybe instead of _tabBlockingState being of type Enum, it is an explicit number representing bit flags instead (0b0000) and you instead of defining the Enum with integer values, you use binary instead?
```ts
enum TabBlocking {
  None = 0,
  Next = 1,
  Previous = 0b10,
  Released = 0b100
}
```

File: packages/main/src/TableVirtualizer.ts [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+		return this._table!.shadowRoot!.getElementById("rows")!;
+	}
+
+	_onScroll(): void {
+		if (!this._table) {
+			return;
+		}
+
+		const scrollTop = this._scrollContainer.scrollTop;
+		const fixHeight = this._table.headerRow[0].offsetHeight;
```
Comment (DonkeyCo) [inline packages/main/src/TableVirtualizer.ts:189]: This is only required if the header is sticky, right? So maybe this should be conditional

File: packages/main/src/TableVirtualizer.ts [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+		this._firstRowPosition = firstRowPosition;
+		this.fireDecoratorEvent<RangeChangeEventDetail>("range-change", {
+			first: firstRowPosition,
+			last: lastRowPosition,
+		});
+	}
+
+	_updateRowsHeight() {
+		const rowsHeight = this.rowCount * this.rowHeight;
+		this._rowsContainer.style.minHeight = `${rowsHeight}px`;
```
Comment (DonkeyCo) [inline packages/main/src/TableVirtualizer.ts:213]: Why minHeight and not height?

File: packages/main/src/TableVirtualizer.ts [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+		this._firstRowPosition = -1;
+		if (this._scrollContainer.scrollTop > 0) {
+			this._scrollContainer.scrollTop = 0;
+		} else {
+			this._onScroll();
+		}
+	}
+
+	get _scrollContainer() {
+		return this._table!._tableElement;
```
Comment (DonkeyCo) [inline packages/main/src/TableVirtualizer.ts:176]: Is this working for something like this?

```html
<div style="height: 375px; overflow: auto">
  <ui5-table>
    <ui5-table-virtualizer row-height="51" row-count="1000" slot="features"></ui5-table-virtualizer>
  </ui5-table>
</div>
```

If that is not the case, we might need to explicitly document the restrictions

File: packages/main/src/TableVirtualizer.ts [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+ * @param {last} number The 0-based index of the last children currently rendered
+ * @public
+ */
+type RangeChangeEventDetail = {
+	first: number,
+	last: number,
+};
+
+/**
+ * @class
```
Comment (DonkeyCo) [inline packages/main/src/TableVirtualizer.ts:43]: It breaks if `overflowMode=Popin`. Maybe a validation would be nice here, that either throws an error or "corrects" the value to `Scroll`
Should be documented as well

File: packages/main/src/TableVirtualizer.ts (hunk 4 above)
Comment (DonkeyCo) [reply packages/main/src/TableVirtualizer.ts:179]: Forget what I wrote, it does not apply anymore

File: packages/main/src/TableVirtualizer.ts [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+	_onScroll(): void {
+		if (!this._table) {
+			return;
+		}
+
+		const headerRow = this._table.headerRow[0];
+		const headerHeight = headerRow.offsetHeight;
+		let scrollTop = this._scrollContainer.scrollTop;
+		let scrollableHeight = this._scrollContainer.clientHeight;
+		if (headerRow.sticky) {
```
Comment (DonkeyCo) [inline packages/main/src/TableVirtualizer.ts:195]: There seems to be an issue with the virtualizer, if the sticky property is added during runtime to the header row:

virtualizer_sticky

I'm fine with ignoring this for now and opening a separate issue to handle this in another PR (just to finally get this feature in, as this case is very very unlikely to happen in an actual app)

**Review by DonkeyCo — COMMENTED** [review]
Looks good to me. Maybe we'll let @KvM2 have a look over the documentation

Otherwise I'm ready to give this a +2. Let me know whether it makes sense to check the docs first

**Comment by aborjinik** [comment]
> Let me know whether it makes sense to check the docs first
would be nice to merge and then i can add the documentation fixes

**Review by DonkeyCo — APPROVED** [review]
LGTM. Docs should be fixed in a later change and an example should be added to the demokit later as well

File: packages/main/src/TableRow.ts [hunk 7]
```diff
@@ -60,6 +60,15 @@ class TableRow extends TableRowBase {
 	@property()
 	rowKey = "";
 
+	/**
+	 * Defines the position of the row respect to the total number of rows within the table when the <code>ui5-table-virtualizer</code> feature is used.
```
Comment (KvM2) [inline packages/main/src/TableRow.ts:64]: in relation to

Comment (KvM2) [inline packages/main/src/TableRow.ts:64]: when -- > if

File: packages/main/src/TableVirtualizer.ts [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+	None = 0,
+	Next = 1,
+	Previous = 2,
+	Released = 4,
+	NextReleased = Next | Released,
+	PreviousReleased = Previous | Released,
+}
+
+/**
+ * Fired when the virtualizer is changed by user interaction e.g. on scrolling.
```
Comment (KvM2) [inline packages/main/src/TableVirtualizer.ts:34]: > e.g. -- >, for example,

File: packages/main/src/TableVirtualizer.ts [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+	first: number,
+	last: number,
+};
+
+/**
+ * @class
+ *
+ * ### Overview
+ *
+ * The `ui5-table-virtualizer` component is used inside the `ui5-table` to virtualize the table rows, if the `overflowMode` property of the table is set to 'Scroll'.
```
Comment (KvM2) [inline packages/main/src/TableVirtualizer.ts:49]: > , if the `overflowMode` property of the table is set to 'Scroll'. no comma

File: packages/main/src/TableVirtualizer.ts [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+ * It is responsible for rendering only the rows that are visible in the viewport and updating them on scroll.
```
Comment (KvM2) [inline packages/main/src/TableVirtualizer.ts:50]: on scrolling

File: packages/main/src/TableVirtualizer.ts [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+ * This allows large numbers of rows to exist, but maintain high performance by only paying the cost for those that are currently visible.
```
Comment (KvM2) [inline packages/main/src/TableVirtualizer.ts:51]: > , but maintain high performance by only paying the cost 
no comma
is pay the cost a common expression ?

File: packages/main/src/TableVirtualizer.ts [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+ *
+ * @constructor
+ * @extends UI5Element
+ * @public
+ * @experimental This component is not intended to be used in a productive enviroment. The API is under development and may be changed in the future.
+ */
+@customElement({ tag: "ui5-table-virtualizer" })
+
+/**
+ * Fired when the virtualizer is changed by user interaction e.g. on scrolling.
```
Comment (KvM2) [inline packages/main/src/TableVirtualizer.ts:66]: > e.g. -- >, for example,

**Comment by KvM2** [comment]
> > Let me know whether it makes sense to check the docs first
> 
> would be nice to merge and then i can add the documentation fixes

I added some comments regarding the language etc . Trying to make this work on Github, let me know if you have a better idea on how to do this.
